Add diffs for created and deleted files

rudolfs opened the pull request:
We'd like to show the file contents for created and deleted files when reviewing code, otherwise we can't do full review within our app.

Here's how it looks in Upstream:
<img width="1728" alt="Screenshot 2022-06-07 at 10 51 15" src="https://user-images.githubusercontent.com/158411/172338754-3de662a4-2021-4fb3-99c1-8ee3309688c7.png">

rudolfs committed:
Add diffs for created and deleted files

Signed-off-by: Rūdolfs Ošiņš <rudolfs@osins.org>

cloudhead commented:
Nice!

rudolfs commented:
> Just need to fix the lints

@geigerzaehler fixed the lints.
cargo-deny is also failing on `master`. 🤷

FintanH commented on surf/src/diff/git.rs:
```diff
@@ -71,14 +71,52 @@ impl<'a> TryFrom<git2::Diff<'a>> for Diff {
                     let path = Path::try_from(path.to_path_buf())?;
 
-                    diff.add_deleted_file(path);
+                    let patch = Patch::from_diff(&git_diff, idx)?;
```
Not a fan of the repeated code here, but I think it needs some reworking with the error types. I might take a crack at splitting it out now.

FintanH replied:
Made the changes here https://github.com/radicle-dev/radicle-surf/pull/204, and fixed the deny along with it

rudolfs replied:
Nice. #240 lgtm. I'll rebase this pr on top of that.

FintanH reviewed (commented):
Mostly looks good! Just a couple of things I'd like addressed, but the overall intention is good :)

FintanH committed:
diff: simplify return type for Diff construction

The return types for constructing Diff were of the form `Result<_,
String>`, however, none of these functions returned an `Err`.

Simplify the types to not use `Result`.

Signed-off-by: Fintan Halpenny <fintan.halpenny@gmail.com>

FintanH committed:
nix: update nix dependencies

Signed-off-by: Fintan Halpenny <fintan.halpenny@gmail.com>

FintanH committed:
deps: update regex >=1.5.5

A vulnerability was found in regex[0]. Pin the dependency to >=1.5.5
as advised.

[0]: https://rustsec.org/advisories/RUSTSEC-2022-0013

Signed-off-by: Fintan Halpenny <fintan.halpenny@gmail.com>

FintanH committed:
diff: introduce a collection of Hunks

To introduce a conversion from git2::Patch to Vec<Hunk> it was
necessary to do two things:
1. Separate out error types to be more granular
2. Introduce a newtype Hunks

1. means that errors are specific to the particular conversions -- and
become nested.

2. was needed for the TryFrom implementation.

Signed-off-by: Fintan Halpenny <fintan.halpenny@gmail.com>

rudolfs commented on surf/examples/diff.rs:
```diff
@@ -122,19 +122,19 @@ fn get_directory_or_exit(browser: &git::Browser) -> Directory {
-        diff.created.len(),
-        diff.deleted.len(),
+        diff.created.len() + diff.created.len(),
+        diff.deleted.len() + diff.deleted.len(),
```
That's just a test. Congrats, you passed! : P

rudolfs replied:
I removed that code in a fixup.
